### ISSUE TITLE
Orchid a11y migration checklist (front end)

### ISSUE THREAD

**Issue by erinchambers** [CDRH/tibetan-monasteries#59]
For overridden files with major changes, feel free to create a separate issue.

# Regex search-and-replace checklist

## Hyphenate classes
- Search with regex: 
``` html
class="(.+)_(.+)"
```
You can replace with `class="$1-$2"`, but be mindful that you do not want to do this across the board.
- Search SCSS with regex: 
``` html
\.(.+)_(.+)\{
```
You can replace with `.$1-$2{`, but be mindful that you do not want to do this across the board. Note in particular that page classes should remain underscores.

## Underscore IDs
- Search with regex: 
``` html
id="(.+)-(.+)"
```
You can replace with `id="$1_$2"`, but be mindful that you do not want to do this across the board. 
- Search SCSS with regex:
``` html
#(.+)-(.+)\{
```
You can replace with `#$1_$2{`, but be mindful that you do not want to do this across the board.

### Files with only hyphenate class/underscore ID changes
- `search.js`: change `search_form` to `search-form`
- `_search_res_items.html.erb`
  - check for multiples of this file (different search presets may have their own) 
- `_sort_and_page_ui.html.erb`
- `search_preset.html.erb`: line 14
- `_languages.html.erb`

# File checklist

### `global/*.scss` files

- [x] make sure there are NO `outline-offset` settings
- hyphenate classes (search with regex, e.g. `\s+\.[a-z]+_`)
- underscore IDs (search with regex, e.g. `\s+#[a-z]+-`)

- [x] Add `Back to Top` link SCSS
``` sass
.back-to-top a {
  background: #fff;
  text-decoration: none;
  font-weight: 500;
  border: solid 1px $gray-light;
  bottom: 0;
  box-shadow: 0px 0px 5px 3px rgba($gray-dark,0.3);
  padding: .5rem 2rem 0.5rem 1.25rem; // .1em 1em .3em 1em; 
  position: fixed;
  right: 2%;
  z-index: 1000;

  &:hover {
  color: #000;
  background-color: $gray-light;
  }

  &:after {
    content: "\2191";
    position: relative;
    top: -1px;
    left: 6px;
  }
}
```
#### Check *.scss for style overrides. What we want to end up with is...
- [x] `#content-wrapper` - Orchid switched to having padding instead of margin on this element. If the site doesn't change anything about #content-wrapper, no action needed.
- [x] underline links by default (if this has unintended fallout, make an issue)
``` scss
a {
text-decoration: underline;
}

a:hover {
text-decoration-thickness: 2px;
}
```
- [x] make sure buttons don't get underlined:
``` sass
a.btn,
a.btn:hover {
text-decoration: none;
}

.btn {
  white-space: initial;
}
```

### `items_controller.rb` and any additional items controllers
- [x] Incorporate code for changing search title, if needed - see [Will's updates](https://github.com/CDRH/orchid/pull/338/files#diff-22da4de4953a8a5748f5a8f4425eb896283d008abd9525a4a33c9fd76f35f40b).

### `display_helper.rb`

- [x] line 10: replace `def metadata(res, label, api_field, link: true, separator: "|", show_label: true)` with (goal here is to remove "|" from separator if it is there)

``` ruby
def metadata(res, label, api_field, link: true, separator: "", show_label: true)
```

- [x] lines 25-29: replace `html << dataArray ...` with

``` ruby
html << "<ul>"
html << dataArray
          .map { |i| "<li>#{i}</li>" }
          .join("")
html << "</ul>"
```

- [x] line 52: replace `Array.wrap...` with

``` ruby
Array.wrap(res[nested.first]).compact.map { |doc| doc[nested[1]] }.flatten
```

## `orchid_setup_generator.rb`

Minor change:
- [x] Correct "footer_logo" typo to "footer logo"

### `_clear_facets.html.erb`

- [x] line 2: add `clear-facets` class ( `<div class="pull-right buffer-bottom-sm clear-facets">`)
- [x] line 7: add sr-only text to glyphicon
```
<span class="glyphicon glyphicon-remove" aria-hidden="true"><span class="sr-only">Remove filter</span>
```

### `_date_limit.html.erb`

- [x] line 20: wrap `<% if date_selection?(... %>` in a new div
```
<div class="clear-date-filter">
	<% if date_selection?(params["date_from"], params["date_to"]) %>
	  <%= link_to prefix_path(route_path, clear_dates_params),
		class: "pull-right btn btn-default btn-xs",
		"aria-label": t("search.dates.clear_filter",
		default: "Clear Date Filter"), rel: "nofollow", role: "button",
		title: "Clear Date Filter" do %>
		<span class="glyphicon glyphicon-remove" aria-hidden="true"></span>
		<% end %>
    <% end %>
</div>
    
```
- [x] change `aria_label` to `"aria-label"`
- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-586ab9a3a2f800297247934b2f138611a5ca8af52b0b64f8b2234d2003f7f505) to `date_from` fields, such as wrapping in fieldsets and adding unique ids

### `_facets.html.erb`

- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-efc8461f37e201619623c18267ecd6e800a13d2d061b4bc25ac0591ad752680e)

### `_help.html.erb`

- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-8b3efdc6384b03094eaa667583509a2534b74a84ae20ef98265d549aaeb0ef5a) - create new issue

### `_index_header.html.erb`

no actual changes (just indentation)

### `_paginator.html.erb`

- [x] line 10: add aria attribute `<nav aria-label="Pagination">`
- [x] line 49: make into a link and add aria attributes `<a href="#" aria-current="page" aria-disabled="true"><%= current %></a>`

### `_search_box.html.erb`

- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-8f82471435236d6fef39a22016b575c18ab2a7eb0659b80fb3068cdb2a745fe1), including columns, order of buttons, and aria roles and labels

### `_search_preset_text`

- [x] wrap in `<div class="container">`
```
<div class="container">
  <%= t "search_preset.about_html" %>
</div>
```

### `_sort.html.erb`

- [x] line 10: add `aria-hidden="true"`
```
<span class="caret" aria-hidden="true"></span>
```

### `_summary_boxes.html.erb`

- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-f68ff4f21a752a7ce8188cd6dfed50a3ba8610b65171e203a7a34a093bb3fd55)

### `browse.html.erb`

- [x] line 10: delete `"aria-controls": id,` 

### `browse_facet.html.erb`

- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-f1bd54707ccdcceb3823001c73cdde380247b7e0427275a5e446680eb76084c3)

### `index.html.erb`
- [ ] wrap `.col-sm-3` and `col-sm-9` in a `<div class="row">`
- [x] Change `col-sm-3` to `col-sm-4 col-md-3` and `col-sm-9` to `col-sm-8 col-md-9`
- #hyphenateclasses  `search_form` -> `search-form`

### `application.html.erb`

- [x] Make sure the `<html>` tag is not missing its closing caret (`>`). It was missing in Orchid. Here's the corrected code:

```
<html lang="<%= I18n.locale %>" <%= sanitize html_classes %>>
```

### `_back_to_top.html.erb`

- if this file already exists on a site (or a Back to Top button is otherwise present), the site had a custom Back to Top button, there may be 2 after the Orchid update and we'll need to pick one.

### `_contents.html.erb`

- [x] convert to `<main>` landmark and delete empty tabindex div
```
<main class="container main-content" id="content_wrapper" tabindex="-1">
	<%= render_overridable("layouts/body", "flash") %>
	<%= yield %>
</main>
```

### `_footer.html.erb`

- [x] Add Back to Top link right before closing tag: 
``` Ruby
<%= render_overridable("layouts/body", "back_to_top") %>
```

- [x] convert to `<footer>` landmark
- #hyphenateclasses 
- [x] change links to `https://`
- [x] add alt text to all images UNLESS a text link to the same info is present, then use `alt="" role="presentation"` so we don't have duplicate links close together

### `_header.html.erb`

- [x] convert to `<header>` landmark

### `_navbar.html.erb`

- [x] [major changes](https://github.com/CDRH/orchid/pull/338/files#diff-a8b3453dc488a3015d66dda0290229b62e88c160563fe101ab98979c8999f652) - create new issue

### `_skip_to_main.html.erb`

- [x] if this file already exists on a site (or a Skip to Main Content link is otherwise present), double-check that things are working after the Orchid upgrade. This may require updating the "skip_to_main" code in `application.html.erb`, or instead (probably preferrable) you can take that code out and add a link to the partial (see [orchid update](https://github.com/CDRH/orchid/commit/54c44e693245b89f447c543d5d603b0d722a5177)).

### `_meta.html.erb`

- [x] Make sure it only has one `<meta charset="..." />` tag. In base Orchid, we removed the shorter of the two that were there, `<meta charset="utf-8"/>`

## YAML
- [x] check for config/sections/*.yml files if something's missing

### `en.yml`
NOTE: it may help to compare to the [updated Orchid file](https://github.com/CDRH/orchid/blob/dev/config/locales/en.yml)
- [x] **Optional:** add UNL link to `cdrh_link_html`
``` yaml
    cdrh_link_html: Created by the <a href="https://cdrh.unl.edu">Center for Digital Research in the Humanities</a> at the <a href="https://www.unl.edu/">University of Nebraska–Lincoln</a>.
```
- [x] line 138: change to `clear_all: Reset Search`
- [x] line 209: add singular `clear: Clear Filter
- [x] lines 224-226: add singular and plural directly underneath`results:`. RETAIN INDENTATION.
``` yaml
      # for calling these labels without numerical values
      singular: result
      plural: results
```
- [x] lines 252-253: add singular and plural directly underneath `results / ... / general:`. RETAIN INDENTATION.
``` yaml
        zero: 0 results
        one: 1 result
        other: "%{count} results"
```
- [x] remove search help html: `<h2 class="sr-only">Search Help</h2>`

### `es.yml`
- [ ] **Optional:** add UNL link to `cdrh_link_html`
``` yaml
    cdrh_link_html: |-
      Creado por el <a href="https://cdrh.unl.edu">Centro de Investigación
      Digital en las Humanidades</a> de la <a href="https://www.unl.edu/">Universidad de Nebraska—Lincoln</a>.
```

- [ ] line 120: change `category: Tipo` to `category2: Tipo`
- [ ] line 209: add singular `clear: Borraro Filtro`
- [ ] lines 224-226: add singular and plural directly underneath`results:`. RETAIN INDENTATION.

``` yaml
      # for calling these labels without the numerical values
      singular: resultado
      plural: resultados
```

- [ ] lines 252-253: add singular and plural directly underneath `results / ... / general:`. RETAIN INDENTATION.

``` yaml
        zero: No se han encontrado resultados para su búsqueda
        one: 1 resultado
        other: "%{count} resultados"
```

**Comment by erinchambers** [comment]
# Bonus Round

###`footer_logo.png`

- [ ] If UNL logo is outdated, upload new N logo OR logo with CDRH name

**Note:** changes to `bootstrap-variables.scss` will be based on site-specific accessibility audit findings, not Orchid changes.

**Comment by erinchambers** [comment]
@wkdewey sorry this task has just been hanging out here. Do you want to switch to the latest Orchid release before launch (or have you already)? If and when we do, this is the checklist I'll go through to make sure things didn't break in the process.

**Comment by wkdewey** [comment]
It does not look like I have updated Orchid, and I think I should (perhaps when we merge your design branch into my branch).

**Comment by erinchambers** [comment]
Cool. Let me know when you plan to do that—I should block off time to make sure I go through the checklist step-by-step.

**Comment by wkdewey** [comment]
I merged the Orchid branch. Should this issue be closed now?

**Comment by erinchambers** [comment]
Yup! Closing.
